### PR TITLE
feat: decode evm errors in traces

### PR DESCRIPTION
**Pull request by meetmangukiya**
Related https://github.com/foundry-rs/foundry/issues/1387

## Motivation

Decode and surface more information in cases of reverts. Many EVM errors can be caught and bubbled up. This PR adds support for decoding these EVM errors.
Other solidity specific error paths need to be inferred by more introspection into evm execution. Yet to be done. 

## Solution

Not-ok `Return` values(evm errors) should be decoded into trace if the call errored and there was no return data(solidity errors).

### CODE REVIEW

**Comment by meetmangukiya** [comment]
Output worth noting for:
- `testStaticCallToNotAStaticCall()` 
- `testLowBalanceTransfer()`
- `testLowBalanceCall()` 

```
[FAIL. Reason: Arithmetic over/underflow] testArithmeticOverflows() (gas: 392)
Traces:
  [189] AnotherTest::setUp() 
    └─ ← ()

  [392] AnotherTest::testArithmeticOverflows() 
    └─ ← "Arithmetic over/underflow"

[PASS] testLowBalanceCall() (gas: 37967)
Traces:
  [37967] AnotherTest::testLowBalanceCall() 
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000001, 1900000000000000000) 
    │   └─ ← ()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000001) 
    │   └─ ← ()
    ├─ [0] 0x0000…000c::fallback{value: 1910000000000000000}() 
    │   └─ ← "EvmError: OutOfFund"
    └─ ← ()

[FAIL. Reason: EvmError: Revert] testLowBalanceTransfer() (gas: 38014)
Traces:
  [189] AnotherTest::setUp() 
    └─ ← ()

  [38014] AnotherTest::testLowBalanceTransfer() 
    ├─ [0] VM::deal(0x0000000000000000000000000000000000000001, 1900000000000000000) 
    │   └─ ← ()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000001) 
    │   └─ ← ()
    ├─ [0] 0x0000…000c::fallback{value: 1910000000000000000}() 
    │   └─ ← "EvmError: OutOfFund"
    └─ ← "EvmError: Revert"

[FAIL. Reason: EvmError: Revert] testNoCode() (gas: 3517)
Traces:
  [189] AnotherTest::setUp() 
    └─ ← ()

  [3517] AnotherTest::testNoCode() 
    ├─ [3000] PRECOMPILE::ecrecover(0dbe671f) 
    │   └─ ← 
    └─ ← "EvmError: Revert"

[PASS] testNonPayableFallback() (gas: 62578)
Traces:
  [62578] AnotherTest::testNonPayableFallback() 
    ├─ [23275] → new A@"0xbc45…da0a"
    │   └─ ← 116 bytes of code
    ├─ [45] A::fallback{value: 1000000000000000000}() 
    │   └─ ← "EvmError: Revert"
    └─ ← ()

[PASS] testRevertingCall() (gas: 1095)
Traces:
  [1095] AnotherTest::testRevertingCall() 
    ├─ [297] AnotherTest::willRevert() 
    │   └─ ← "SomeError()"
    └─ ← ()

[PASS] testStaticCallToNotAStaticCall() (gas: 9079256848778899261)
Traces:
  [9079256848778899261] AnotherTest::testStaticCallToNotAStaticCall() 
    ├─ [2392] AnotherTest::notAStaticCall() [staticcall]
    │   └─ ← "EvmError: NotActivated"
    └─ ← ()

Test result: FAILED. 4 passed; 3 failed; finished in 884.38µs

Failed tests:
[FAIL. Reason: Arithmetic over/underflow] testArithmeticOverflows() (gas: 392)
[FAIL. Reason: EvmError: Revert] testLowBalanceTransfer() (gas: 38014)
[FAIL. Reason: EvmError: Revert] testNoCode() (gas: 3517)
```

for 

```solidity
contract A {
    function a() external returns (uint256) {
        return 1;
    }

    fallback() external {}
}

contract DecodeRevertsTest is Test {
    address code = address(1);
    error SomeError();
    uint256 counter = 0;

    function willRevert() public {
        revert SomeError();
    }

    function testRevertingCall() public {
        address(this).call(abi.encodeWithSignature("willRevert()"));
    }

    function notAStaticCall() public {
        counter += 1;
    }

    function testStaticCallToNotAStaticCall() public {
        address(this).staticcall(abi.encodeWithSignature("notAStaticCall()"));
    }

    function testNonPayableFallback() external {
        A a = new A();
        (bool success, bytes memory ret) = address(a).call{value: 1 ether}("");
    }

    function testNoCode() external {
        A(address(1)).a();
    }

    function testLowBalanceTransfer() external {
        deal(address(1), 1.9 ether);
        vm.prank(address(1));
        payable(address(12)).transfer(1.91 ether);
    }

    function testLowBalanceCall() external {
        deal(address(1), 1.9 ether);
        vm.prank(address(1));
        address(12).call{value: 1.91 ether}("");
    }

    function testArithmeticOverflows() external {
        uint256 i = 1;
        i -= 1;
        i -= 1;
    }
}
```

**Comment by onbjerg** [comment]
As a general thought, why do we have a separate "status" field instead of catching all of the revert reasons in `result` in REVM?

**Comment by onbjerg** [comment]
Also related: #476

**Comment by meetmangukiya** [comment]
> As a general thought, why do we have a separate "status" field instead of catching all of the revert reasons in `result` in REVM?

`RawCallResult` had a status field earlier too, made it public within crate. I am guessing it is just so to not lose any data that we are getting from revm? The `result` has the return value from the call. The same in `CallResult` is a decoded value or an `EvmError::Execution` is returned if the call was a reverting one.

**Comment by mattsse** [comment]
any update on this?

what's missing?

**Comment by mattsse** [comment]
moved `decode_revert` to evm

this now properly decodes all new rpc/fork errors:

image

File: evm/src/decode.rs [hunk 1]
```diff
@@ -52,3 +62,142 @@ pub fn decode_console_log(log: &Log) -> Option<String> {
...
+                .or_else(|| {
+                    // try decoding as unknown err
+                    String::decode(&err[SELECTOR_LEN..])
+                        .map(|err| format!("{}:{}", hex::encode(&err[..SELECTOR_LEN]), err))
+                        .ok()
+                })
+                .ok_or_else(|| eyre::eyre!("Non-native error and not string"))
+
+            // evm_error will sometimes not include the function selector for the error,
+            // optimistically try to decode
```
Comment (onbjerg) [inline evm/src/decode.rs:200]: nit: unrelated cleanup

```suggestion

```

**Review by onbjerg — APPROVED** [review]
lgtm
